File: 159. Longest Substring with At Most Two Distinct Characters/solution.py
class Solution(object):
    def lengthOfLongestSubstringTwoDistinct(self, s):
        """
        :type s: str
        :rtype: int
        """
        unique = {}
        i = 0
        j = 0
        l = 0
        while j < len(s):
            if s[j] not in unique:
                unique[s[j]] = 0
            unique[s[j]] += 1
            j += 1
            if len(unique) > 2:
                unique[s[i]] -= 1
                if not unique[s[i]]:
                    unique.pop(s[i])
                    
                i += 1
                
            l = max(l, j-i)
            
        return l

File: 159. Longest Substring with At Most Two Distinct Characters/solution.java
class Solution {
    public int lengthOfLongestSubstringTwoDistinct(String s) {
        HashSet<Character> bucket = new HashSet<>();
        Character lastChar = null;
        Character c;
        int lastCharLen = 0;
        int len = 0;
        int maxLen = 0;
        int i = 0;
        while (i < s.length()) {
            c = s.charAt(i);
            if (bucket.size() < 2) {
                if (bucket.contains(c)) {
                    lastCharLen++;
                } else {
                    lastCharLen = 1;
                }
                lastChar = c;
                bucket.add(c);
            } else {
                if (bucket.contains(c)) {
                    if (lastChar == c) {
                        lastCharLen++;
                    } else {
                        lastChar = c;
                        lastCharLen = 1;
                    }
                } else {
                    bucket.clear();
                    bucket.add(lastChar);
                    bucket.add(c);
                    len = lastCharLen;
                    lastCharLen = 1;
                    lastChar = c;
                }
            }
            len++;
            maxLen = Math.max(len, maxLen);
            i++;
        }
        return maxLen;
    }
}
